controllers/progressiverollout_controller.go

```diff
@@ -67,6 +70,32 @@ func (r *ProgressiveRolloutReconciler) Reconcile(ctx context.Context, req ctrl.R
 
 	log = r.Log.WithValues("applicationset", pr.Spec.SourceRef.Name)
 
+	if pr.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object
+		if !controllerutil.ContainsFinalizer(&pr, deploymentskyscannernetv1alpha1.ProgressiveRolloutFinalizer) {
+			controllerutil.AddFinalizer(&pr, deploymentskyscannernetv1alpha1.ProgressiveRolloutFinalizer)
+			if err := r.Update(ctx, &pr); err != nil {
+				r.Log.Error(err, "failed to update object")
+				return ctrl.Result{}, err
+			}
+			// Requeue after adding the finalizer
```

Comment:
Can you expand on this? I'm unclear why the requeue is set, given the finaliser is added once and stays until the object is deleted? Am i missing some kind of race condition?

Reply:
They don't do this in the docs.

As on every update and create `DeletionTimestamp.IsZero()` will always be true, does this mean that we never get past this point in the code?

Reply:
Yeah when I checked the doc I was confused too.
I guess it doesn't make much difference but uses a full reconcile loop just to add the finalizer.
@maruina did you want to trigger the loop as fast as possible to avoid some kind of condition?

Reply:
I know they don't do that in the book but I think it's still a good idea because it tries to keep the loop as small as possible in the case of a controller crash. Instead of adding a finalizer AND doing the first stage, we only add the finalizer. 

If we decide this is an unnecessary optimization, we should at least change this part to use https://github.com/Skyscanner/argocd-progressive-rollout/pull/60/files#diff-b1a5cf22f2be07477047c7c268c9a2077f75c3fdadc285eb862dab8bc3aaad31R400 to avoid updating an old object.